### ISSUE TITLE
File I/O failing with errno=4 (EINTR)

### ISSUE THREAD

**Issue by dmcnaugh** [udo-munk/z80pack#127]
While I've been working on the WDI-II emulation I've been getting occasional file I/O failures.
These have happened with the hard disk image files but also with floppy disk images files.

Some investigation has lead to seeing the file `open()` is failing with an `errno` of 4 - `EINTR`
This indicates that in interrupt has occurred during the file `open()`, highly possible as interrupts/signals are used with the _cromemcosim_ for the emulated interrupts and timers.

Some background reading suggests that when the signal is initialised the `sa_flags` attribute can be set to `SA_RESTART` to prevent this behaviour (this is the default behaviour in BSD apparently, but not POSIX).

I tried doing this everywhere but continued to get occasional `EINTR` errors returned while opening disk files anyway.

A retry loop is required around the file `open()` until the `EINTR` condition no longer occurs.

I will include this in the `cromemco-wdi.c` code (next PR), but it should also be done in the various floppy disk emulators too.

**Comment by udo-munk** [comment]
On UNIX (like) systems disk I/O cannot be interrupted, this would cause endless problems with inconsistent data. This is why you see handling of interrupts only in the tty/terminal routines and not on any disk I/O. I have not yet seen any interrupted disk I/O on any UNIX (like) system with the emulations and such an unhandled interrupt would abort every program often, so I have some doubt about this issue.
Of course you can try to handle this in the WTI code, but please don't change any FDC code yet.

**Comment by dmcnaugh** [comment]
I don't intend on changing any of the fdc code. I raised this as an issue so that is can be discussed ahead of any solution if ultimately required.

You are correct, disk I/O cannot be interrupted - `read()` and `write()` and related functions. But it appears that `open()` can be interrupted. I have adopted the same strategy for the _WDI_ as for the _fdc_'s: ie. the host file is opened and closed for each read or write.

So I am seeing the `errno == EINTR` returned when the `open()` returns with `-1`.

**Comment by udo-munk** [comment]
open() can be interrupted for slow charcater devices, but won't get interrupted on fast block devices.
All emulations other then cpmsim open the disk, do read/write and close it, cpmsim keeps the files open for somewhat faster performance. I have not seen yet on any OS used that an open() of a disk file get's interrupted and I think that this never should happen.

**Comment by dmcnaugh** [comment]
Maybe this is a bug in macOS 12.4 Monterey on M1 (Mac mini)? I will test on my i5 macbook pro and see if it also occurs.

But it is happening. I wish it wasn't, but it is.

Here is the code and the output log that demonstrates the problem, this is in the disk write function:

```
    int got_eintr = 0;
again:
    if ((fd = open(fn, O_RDONLY)) == -1) {
        if (errno == EINTR) {
            if (!got_eintr) LOGW(TAG, "READ: GOT EINTR - %s : errno %d", fn, errno);
            got_eintr++;
            goto again;
        } else {
            LOGE(TAG, "READ: HDD FILE DOES NOT EXIST - %s : %s [%d]", fn, strerror(errno), errno);
            wdi.hd[unit]._fault = 0; /* SET FAULT */
            close(fd);
            return 0;
        }
    }
    if (got_eintr) LOGW(TAG, "READ: GOT EINTR: total %d", got_eintr);
```

```
W (1924203369) wdi: READ: GOT EINTR - ./disks/hd0.hdd : errno 4
W (1924505850) wdi: READ: GOT EINTR: total 375
```

In this example the delay was around 1 second. I have seen the loop counter range from < 100 to > 3000.
When the loop counter is around 3000 the delay is in the order of 5 or more seconds.

**Comment by ratboy666** [comment]
Yes, EINTR can be raised on open() , openat() and creat(). This is not
a bug in macOS.

This will happen if, for example, alarm() is used.

close() can also raise EINTR -- but do not retry in that case
(folklore, mostly)

FredW

**Comment by EtchedPixels** [comment]
close can cause EINTR and not retrying can lose bytes on some NFS implementations, it also leaks the handle if you don't retry 8)

The best approach is probably to use sigaction() and restart. Apart from old Sys5 unix nothing should lack that.

**Comment by dmcnaugh** [comment]
> The best approach is probably to use sigaction() and restart. Apart from old Sys5 unix nothing should lack that.

As a test, I went through the whole Z80PACK code base and set `xxx.sa_flags = SA_RESTART` for all `sigaction()` calls and still continued to get occasional `EINTR` on the disk file `open()` calls. Very frustrating.

Maybe I missed something, so I will try again.

Meanwhile, I have a workaround solution: only `open()` the hard disk file **once** on initialisation; use `rsync()` after every `write()` to flush data as much as possible; only `close()` the files on reset or exit. This is probably the long term solution for the WDI-II implementation anyway. But, `EINTR` can and still does occur with the FDC's when they `open()` and `close()` files on every read/write.

**Comment by EtchedPixels** [comment]
open/close do not flush the data to media anyway so there's really no difference between open write close and opening the file and leaving it open all day except over NFS in some cases, and nobody uses NFS any more do they ?

In other words you don't need the sync, in fact you really really don't want to be doing lots of sync style operations on modern storage.

**Comment by udo-munk** [comment]
If even the SA_RESTART flag doesn't work they must have screwed up the signal handling in the OS a bit. I'm still on OSX 10.9.5 and cannot reproduce the problem, just doesn't happen.

If close() aborts with errno EINTR and you don't repeat it, it probably will not free the fd and it gets lost. You can do this so often until you run into the ulimit for open fd's for the process, so one won't notice immediatley that something is going wrong.

There is one huge difference between keeping disk files open, or open/close them for every access. The later allows to change a disk image while emulation is running and you need that in some situations for the floppy disks.

Sure do I use NFS, wherever on the planet I login to any machine of some organsition I'll have my home directory automounted from an NFS server. Also on that systems I never got errno EINTR so far.
After vacation I will run a VM with Cromix on HD for 24 hours or so, running some scripts that causes lots of disk activities.

**Comment by dmcnaugh** [comment]
I have been unable to reproduce this problem on my Intel i5 MacBook Pro.

After more than a week of being able to reliably reproduce the problem on my M1 Mac Mini:
- I did not reboot
- I did not install any updates (that I know of)
- It . just .. stopped ... happening!
- and I can no longer reproduce the problem - I hate these kind of phantom issues.

I've decided to proceed with the strategy of opening the hard disk file(s) only once and keeping them open until a shutdown or reset (Ext.Clr.) as these are not removable media like a floppy.
I've also included the retry loop on `EINTR` for `open()` - just in case.

So with this issue out of the way, I am just cleaning up the harddisk representation on the webfrontend and then I will post a PR with these changes.

Sorry for the annoyance. I too, wish it had not happened.
